File: c/ceres/source/ceres/iterable/SortedList.c
/*
 * The MIT License (MIT)
 *
 * Copyright © 2013-2018 Florian Barras <https://barras.io> (florian@barras.io)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
/***************************************************************************************************
 * INCLUDES
 **************************************************************************************************/

#include "ceres/iterable/SortedList.h"


/***************************************************************************************************
 * CONSTRUCT
 **************************************************************************************************/

SortedList* SortedList_new(void)
{
	SortedList* sl = _NEW(SortedList);

	_PRINT_DEBUG(_S("<newSortedList>"));
	if (sl != NULL)
	{
		sl->core = Core_create(_TRUE, _FALSE, _TRUE, _TRUE);
		sl->length = 0;
		sl->size = 0;
		sl->first = NULL;
		sl->last = NULL;
		SortedList_reset(sl);
	}
	else
	{
		_PRINT_ERROR_MEMORY_ALLOCATION(_SORTED_LIST_NAME);
	}
	_PRINT_DEBUG(_S("</newSortedList>"));
	return sl;
}


/***************************************************************************************************
 * RESET
 **************************************************************************************************/

void SortedList_reset(void* sortedList)
{
	_PRINT_DEBUG(_S("<resetSortedList>"));
	_IF (_CHECK(sortedList, _SORTED_LIST_NAME))
	{
		/* Get the Sorted List */
		SortedList* sl = (SortedList*) sortedList;

		/* INHERITANCE */
		LinkedList_reset(sl);

		/* FUNCTIONS */
		/* - Collection */
		sl->add = SortedList_add;
		sl->addValue = SortedList_add_value;
		sl->addStructure = SortedList_add_Structure;
	}
	_PRINT_DEBUG(_S("</resetSortedList>"));
}


/***************************************************************************************************
 * SORTED LIST
 **************************************************************************************************/

boolean SortedList_insert(SortedList* sortedList, SortedNode* node)
{
	_IF (_CHECK(sortedList, _SORTED_LIST_NAME)
		&& _CHECK(node, _SORTED_NODE_NAME))
	{
		if (sortedList->first == NULL)
		{
			_PRINT_DEBUG(_S("first and last"));
			sortedList->first = node;
			sortedList->last = node;
		}
		else if (node->compareTo(node, _OBJECT_TYPE, sortedList->last) >= 0)
		{
			_PRINT_DEBUG(_S("last"));
			/* ... o->o NULL */
			sortedList->last->next = node;
			/* ... o<-o NULL */
			node->previous = sortedList->last;
			/* Set the last node */
			sortedList->last = node;
		}
		else
		{
			/* Get the first Sorted Node */
			SortedNode* current = sortedList->first;

			do
			{
				if (node->compareTo(node, _OBJECT_TYPE, current) <= 0)
				{
					SortedNode* previous;

					previous = current->previous;
					if (previous == NULL)
					{
						_PRINT_DEBUG(_S("first"));
						/* NULL o->o ... */
						node->next = current;
						/* NULL o<-o ... */
						current->previous = node;
						/* Set the first node */
						sortedList->first = node;
					}
					else
					{
						_PRINT_DEBUG(_S("inside"));
						/* ... o->o o ... */
						previous->next = node;
						/* ... o<-o o ... */
						node->previous = previous;
						/* ... o o->o ... */
						node->next = current;
						/* ... o o<-o ... */
						current->previous = node;
					}
					break;
				}
				current = current->next;
			}
			while (_TRUE);
		}
		++sortedList->length;
		sortedList->size = sortedList->length;
		return _TRUE;
	}
	return _FALSE;
}


/***************************************************************************************************
 * COLLECTION
 **************************************************************************************************/

boolean SortedList_add(void* collection, const type type, void* value)
{
	_IF (_CHECK(collection, _SORTED_LIST_NAME)
		&& _CHECK(value, _S("value")))
	{
		/* Get the Sorted List */
		SortedList* sl = (SortedList*) collection;
		/* Declare the Sorted Node */
		SortedNode* node;

		/* Create the Sorted Node to be inserted */
		if (type == _OBJECT_TYPE)
		{
			node = LinkedNode_new_from_Object((Object*) value);
		}
		else if (type == _STRUCTURE_TYPE)
		{
			node = LinkedNode_new_from_Structure((Structure*) value);
		}
		else
		{
			const Structure s = Structure_create(type, value);

			node = LinkedNode_new_from_Structure(&s);
		}
		return SortedList_insert(sl, node);
	}
	return _FALSE;
}

boolean SortedList_add_value(void* collection, void* value)
{
	_IF (_CHECK(collection, _SORTED_LIST_NAME)
		&& _CHECK(value, _S("value")))
	{
		/* Get the Sorted List */
		SortedList* sl = (SortedList*) collection;
		/* Create the Sorted Node to be inserted */
		SortedNode* node = LinkedNode_new_from_Object(value);

		return SortedList_insert(sl, node);
	}
	return _FALSE;
}

boolean SortedList_add_Structure(void* collection, const Structure* structure)
{
	_IF (_CHECK(collection, _SORTED_LIST_NAME)
		&& _STRUCTURE_CHECK(structure))
	{
		/* Get the Sorted List */
		SortedList* sl = (SortedList*) collection;
		/* Create the Sorted Node to be inserted */
		SortedNode* node = LinkedNode_new_from_Structure(structure);

		return SortedList_insert(sl, node);
	}
	return _FALSE;
}

File: java/jupiter/common/source/jupiter/common/struct/list/SortedList.java
/*
 * The MIT License (MIT)
 *
 * Copyright © 2013-2020 Florian Barras <https://barras.io> (florian@barras.io)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package jupiter.common.struct.list;

import java.util.Collection;
import java.util.LinkedList;

import jupiter.common.model.ICloneable;
import jupiter.common.test.CollectionArguments;
import jupiter.common.util.Collections;
import jupiter.common.util.Integers;

/**
 * {@link SortedList} extends {@link LinkedList} of element type {@code E}.
 * <p>
 * @param <E> the self element {@link Comparable} type of the {@link SortedList}
 */
public class SortedList<E extends Comparable<E>>
		extends LinkedList<E>
		implements ICloneable<SortedList<E>> {

	////////////////////////////////////////////////////////////////////////////////////////////////
	// CONSTANTS
	////////////////////////////////////////////////////////////////////////////////////////////////

	/**
	 * The generated serial version ID.
	 */
	private static final long serialVersionUID = 1L;


	////////////////////////////////////////////////////////////////////////////////////////////////
	// CONSTRUCTORS
	////////////////////////////////////////////////////////////////////////////////////////////////

	/**
	 * Constructs an empty {@link SortedList} of element type {@code E}.
	 */
	public SortedList() {
		super();
	}

	/**
	 * Constructs a {@link SortedList} of element type {@code E} with the specified
	 * {@link Collection} of element subtype {@code E}.
	 * <p>
	 * @param collection a {@link Collection} of element subtype {@code E}
	 * <p>
	 * @throws NullPointerException if {@code collection} is {@code null}
	 */
	public SortedList(final Collection<? extends E> collection) {
		super(collection);
	}


	////////////////////////////////////////////////////////////////////////////////////////////////
	// GETTERS
	////////////////////////////////////////////////////////////////////////////////////////////////

	/**
	 * Returns the element {@link Class}.
	 * <p>
	 * @return the element {@link Class}
	 */
	public Class<E> getElementClass() {
		return Collections.getElementClass(this);
	}

	public E getMiddle() {
		// Check the arguments
		CollectionArguments.requireNotEmpty(this);

		// Get the middle
		return get(Integers.middle(size()));
	}


	////////////////////////////////////////////////////////////////////////////////////////////////
	// SETTERS
	////////////////////////////////////////////////////////////////////////////////////////////////

	@Override
	public synchronized E set(final int index, final E element) {
		return super.set(index, element);
	}


	////////////////////////////////////////////////////////////////////////////////////////////////
	// CONVERTERS
	////////////////////////////////////////////////////////////////////////////////////////////////

	/**
	 * Returns an array containing all of the elements in {@code this} in proper sequence (from
	 * first to last element), or an empty array if {@code this} is empty.
	 * <p>
	 * @return an array containing all of the elements in {@code this} in proper sequence (from
	 *         first to last element), or an empty array if {@code this} is empty
	 *
	 * @see LinkedList#toArray
	 */
	@Override
	public Object[] toArray() {
		return Collections.toArray(this);
	}

	/**
	 * Returns a primitive array containing all of the elements in {@code this} in proper sequence
	 * (from first to last element), or {@code null} if {@code this} is empty.
	 * <p>
	 * @return a primitive array containing all of the elements in {@code this} in proper sequence
	 *         (from first to last element), or {@code null} if {@code this} is empty
	 *
	 * @see LinkedList#toArray
	 */
	public Object toPrimitiveArray() {
		return Collections.toPrimitiveArray(this);
	}


	////////////////////////////////////////////////////////////////////////////////////////////////
	// FUNCTIONS
	////////////////////////////////////////////////////////////////////////////////////////////////

	@Override
	public synchronized boolean add(final E element) {
		int index = 0;
		for (final E e : this) {
			if (e.compareTo(element) > 0) {
				add(index, element);
				return true;
			}
			++index;
		}
		super.add(element);
		return true;
	}

	@Override
	public synchronized void add(final int index, final E element) {
		super.add(index, element);
	}

	public synchronized <T extends E> boolean addAll(final T[] array) {
		for (final E element : array) {
			add(element);
		}
		return true;
	}

	@Override
	public synchronized boolean addAll(final Collection<? extends E> collection) {
		for (final E element : collection) {
			add(element);
		}
		return true;
	}

	@Override
	public synchronized boolean addAll(final int index, final Collection<? extends E> collection) {
		return addAll(collection);
	}

	////////////////////////////////////////////////////////////////////////////////////////////////

	@Override
	public synchronized E remove(final int index) {
		return super.remove(index);
	}

	@Override
	public synchronized boolean remove(final Object object) {
		return super.remove(object);
	}

	@Override
	public synchronized boolean removeAll(final Collection<?> collection) {
		return super.removeAll(collection);
	}

	////////////////////////////////////////////////////////////////////////////////////////////////

	@Override
	public synchronized boolean retainAll(final Collection<?> collection) {
		return super.retainAll(collection);
	}

	////////////////////////////////////////////////////////////////////////////////////////////////

	@Override
	public synchronized void clear() {
		super.clear();
	}


	////////////////////////////////////////////////////////////////////////////////////////////////
	// OBJECT
	////////////////////////////////////////////////////////////////////////////////////////////////

	/**
	 * Creates a copy of {@code this}.
	 * <p>
	 * @return a copy of {@code this}
	 *
	 * @see jupiter.common.model.ICloneable
	 */
	@Override
	@SuppressWarnings("unchecked")
	public SortedList<E> clone() {
		return (SortedList<E>) super.clone();
	}

	////////////////////////////////////////////////////////////////////////////////////////////////

	/**
	 * Returns a representative {@link String} of {@code this}.
	 * <p>
	 * @return a representative {@link String} of {@code this}
	 */
	@Override
	public String toString() {
		return Collections.toString(this);
	}
}
